interfaces/builtin/gpg_keys.go

```diff
@@ -49,6 +49,15 @@ deny @{HOME}/.gnupg/trustdb.gpg w,
 # trust to access the user's keys and the level of trust for the random_seed
 # is commensurate with accessing the private keys.
 owner @{HOME}/.gnupg/random_seed wk,
+
+# allow access to gpg-agent's extra socket for passphrase entry.
+# This socket exposes a very limited subset of gpg-agent's functionality and
+# only allows the confined application to use gpg's cryptographic functions
+# but not anything related to managing the keys (adding, trusting etc).
+owner /run/user/[0-9]*/gnupg/S.gpg-agent.extra wr,
+
+# gpg creates lockfiles on every invocation
+owner @{HOME}/.gnupg/.#lk* rwk,
```

Comment:
This was not required for `/usr/bin/gpg{,1,2,v}` in the past when not using the agent. What is creating these lockfiles? Can you provide exact steps on how to reproduce (when not using the agent)? `snap run --strace ...` would likely help shed some light.

Reply:
> This was not required for `/usr/bin/gpg{,1,2,v}` in the past when not using the agent. What is creating these lockfiles? Can you provide exact steps on how to reproduce (when not using the agent)? `snap run --strace ...` would likely help shed some light.

I looked at this and found that the core18 base snap only contains gpgv. So I copied gpg and various libs it needed to SNAP_COMMON of a snap with this interface connected and did:
```
LD_LIBRARY_PATH=$SNAP_COMMON GNUPGHOME=/home/jamie/.gnupg $SNAP_COMMON/gpg --list-keys --no-auto-check-trustdb
...
uid           [ unknown] Test Foo <test.foo@example.com>
...

$ LD_LIBRARY_PATH=$SNAP_COMMON GNUPGHOME=/home/jamie/.gnupg $SNAP_COMMON/gpg --sign foo
...
gpg: failed to create temporary file '.../.gnupg/.#lk0x00005603e517cda0.sec-bionic-amd64.6029': Permission denied
```

As such, if this rule were to be added, I think it should be:
```
# required for talking with agent
owner @{HOME}/.gnupg/.#lk0x[0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f][0-9a-f].*.[0-9]* rw,
```

*but* this exercise reminded me that gpg v2 *always* must be able to talk to the agent. Per the man page:
```
       --use-agent
       --no-use-agent
              This is dummy option. gpg always requires the agent.
```

As such, there is no point adding these lockfiles to this interface since they are for use with the agent. The above rule would need to be in the new agent interface, if it was created.

As mentioned, none of this is needed by consumers of this interface. I [created a test-gnupg2 snap](https://code.launchpad.net/~jdstrand/+git/test-gnupg2) (it is not production; obtaining a tty is non-trivial and so there are limitations) and can demonstrate that everything works[1]:
```
$ test-gnupg2.gpg --version
gpg (GnuPG) 2.2.4
libgcrypt 1.8.1
Copyright (C) 2017 Free Software Foundation, Inc.
...
$ test-gnupg2.gpgconf --list-dirs
...
socketdir:/home/ubuntu/snap/test-gnupg2/current/.gnupg
...
agent-socket:/home/ubuntu/snap/test-gnupg2/current/.gnupg/S.gpg-agent
homedir:/home/ubuntu/snap/test-gnupg2/current/.gnupg

# generate a key in ~/.gnupg on the host, outside of the snap
$ gpg --quick-gen-key --yes host.key@example.com
...
uid                      host.key@example.com
...

# now do various operations within the snap
$ test-gnupg2.gpg --list-keys
gpg: keybox '/home/ubuntu/snap/test-gnupg2/current/.gnupg/pubring.kbx' created
gpg: /home/ubuntu/snap/test-gnupg2/current/.gnupg/trustdb.gpg: trustdb created

$ test-gnupg2.gpg --quick-gen-key --yes test.me@example.com
...
gpg: key 5C38153FCF4E759B marked as ultimately trusted
gpg: directory '/home/ubuntu/snap/test-gnupg2/current/.gnupg/openpgp-revocs.d' created
gpg: revocation certificate stored as '/home/ubuntu/snap/test-gnupg2/current/.gnupg/openpgp-revocs.d/85384036A5605D16004490295C38153FCF4E759B.rev'
...
uid                      test.me@example.com

$ echo testme > ~/snap/test-gnupg2/common/foo

$ cd ~/snap/test-gnupg2/common ; test-gnupg2.gpg -r test.me@example.com -e ~/snap/test-gnupg2/common/foo
gpg: checking the trustdb
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
gpg: next trustdb check due at 2021-09-04

$ cd ~/snap/test-gnupg2/common ; test-gnupg2.gpg -d ~/snap/test-gnupg2/common/foo.gpg 
gpg: encrypted with 3072-bit RSA key, ID 4FCA2D2BAF62CAE9, created 2019-09-05
      "test.me@example.com"
testme

# now try to use the key outside the snap
$ cd ~/snap/test-gnupg2/common ; test-gnupg2.gpg -r host.key@example.com -e ~/snap/test-gnupg2/common/foo
...
gpg: error retrieving 'host.key@example.com' ...
...

$ test-gnupg2.import-from-host
Imported /home/ubuntu/.gnupg to /home/ubuntu/snap/test-gnupg2/x8/.gnupg.host
Use --homedir=~/.gnupg.host for these keys

$ cd ~/snap/test-gnupg2/common ; test-gnupg2.gpg --homedir=~/.gnupg.host -r host.key@example.com -e ~/snap/test-gnupg2/common/foo # gpg expands '~' to $HOME, ie, $SNAP_USER_DATA
gpg: checking the trustdb
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
gpg: next trustdb check due at 2021-09-04
File '/home/ubuntu/snap/test-gnupg2/common/foo.gpg' exists. Overwrite? (y/N) y

$ cd ~/snap/test-gnupg2/common ; test-gnupg2.gpg --homedir=~/.gnupg.host -d ~/snap/test-gnupg2/common/foo.gpg 
gpg: encrypted with 3072-bit RSA key, ID A561E82BF3C5C438, created 2019-09-05
      "host.key@example.com"
testme
```

[1]gpg unconditionally tries to create the /run/user/id/gnupg/d.... directory but falls back to GNUPGHOME if it is unable to, so there is a spurious AppArmor noisy denial: `type=1400 audit(1567701445.512:1673): apparmor="DENIED" operation="mkdir" profile="snap.test-gnupg2.gpg" name="/run/user/1001/gnupg/d.tbfxb9f6by7d4myapk8qbyud/" pid=11740 comm="gpg-agent" requested_mask="c" denied_mask="c" fsuid=1001 ouid=1001`